### PR TITLE
feat(connector): send metadata in payment authorize request for noon nmi cryptopay

### PR DESCRIPTION
**Pull request by hrithikesh026**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Send metadata in the authorize request for noon, nmi and cryptopay.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Manual.
1. CryptoPay (connector force sync response)
Cryptopay log

2. NMI (connector force sync response)
Authomatic capture (auth+capture)
Screenshot 2024-02-06 at 12 36 19 PM

Manual Capture(auth and then capture)
Screenshot 2024-02-06 at 1 16 28 PM

4. Noon (connector dashboard)
Screenshot 2024-01-31 at 8 25 10 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/noon/transformers.rs [hunk 1]
```diff
@@ -67,9 +67,46 @@ pub struct NoonOrder {
...
+    inner: std::collections::BTreeMap<String, Secret<String>>,
+}
+
+fn get_value_as_string(value: &serde_json::Value) -> String {
+    match value {
+        serde_json::Value::Null => "null".to_string(),
+        serde_json::Value::Bool(bool) => bool.to_string(),
+        serde_json::Value::Number(num) => num.to_string(),
+        serde_json::Value::String(string) => string.to_owned(),
+        serde_json::Value::Array(_) => "null".to_string(),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/noon/transformers.rs:86]: Ideally merchant should not pass any value as object in the metadata, But if he passes then let's not modify the merchant data to null, Instead pass whatever being passed to us to the connector and let the payment fail at the connector

File: crates/router/src/connector/nmi/transformers.rs [hunk 2]
```diff
@@ -451,6 +477,12 @@ impl TryFrom<&NmiRouterData<&types::PaymentsAuthorizeRouterData>> for NmiPayment
             amount,
             currency: item.router_data.request.currency,
             payment_method,
+            merchant_defined_field: item
```
[the remark this answers is not in the document]
Comment (hrithikesh026) [reply crates/router/src/connector/nmi/transformers.rs:480]: metadata is available in capture and void transaction.

File: crates/router/src/connector/nmi/transformers.rs [hunk 3]
```diff
@@ -564,6 +596,7 @@ impl TryFrom<&types::SetupMandateRouterData> for NmiPaymentsRequest {
             amount: 0.0,
             currency: item.request.currency,
             payment_method,
+            merchant_defined_field: None,
```
[the remark this answers is not in the document]
Comment (hrithikesh026) [reply crates/router/src/connector/nmi/transformers.rs:599]: setup mandate flow is not implemeted for Cryptopay, nmi and noon.

**Comment by ArjunKarthik** [comment]
@hrithikesh026 For each connector please add test case for Manual capture, void, where metadata passed in Authorize flow should be visible in all the transaction (Authorize, Capture, Void) in the connector dashboard.

**Review by ArjunKarthik — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Pull request merged**